### PR TITLE
add intersection's solution

### LINKED ISSUE #3: Intersection
## Behavior

> What the function does - inputs and outputs

- **The Function**
  - [x] The function's name is clear and helpful
  - [x] The parameter names are clear and helpful
  - [x] Default parameters are used if possible
- **JSDoc Comment**
  - [x] The docstring is well-formatted
  - [x] There are no spelling mistakes
  - [x] The description of the function's behavior is clear
  - [x] All parameters are documented with types
  - [x] Documented parameter names match the function's parameter names
  - [x] Parameters have a good description
  - [x] If default parameters are used, they are the correct type
  - [x] The return value is documented with a type
  - [x] The return value has a clear description
- **README**
  - [x] The documentation is rendered into the README
  - [x] Docs were generated just before pushing
  - [x] Docs in the README matches the JSDoc comment
  - [x] The README contains at least one use case
- **Unit Tests**
  - [x] Unit tests are present
  - [x] There are at least 8 test cases
  - [x] All `describe`s and `it`s have clear and helpful names
  - [x] All test names actually describe what is being tested
  - [x] Tests are grouped logically using `describe`
  - [x] There is only one `expect` inside every `it`
  - [x] There is no logic (loops, conditionals, ...) in the test cases
  - [x] Side-effects are tested, if necessary
  - [x] Tests were run just before pushing
  - [x] All tests pass (`file-name.pass.txt`)
  - [x] Tricky edge cases are tested (extra credit)

## Strategy

> The logic behind the solution

- **The Function**
  - [x] Variables have names that help understand the strategy
  - [x] Spacing and formatting help to understand the strategy
  - [x] The code is written as simply and clearly as possible
  - [x] Comments _are_ used to label important goals in your strategy
  - [x] Comments _are not_ used to describe how the code works
- **README**
  - [x] There is an explanation of the solution's strategy in the README
  - [x] The explanation is clear and understandable
  - [x] The explanation actually matches the function that was written

## Implementation

> The code used to write the solution

- **Formatting**
  - [x] The code was formatted just before pushing
  - [x] No errors occurred while formatting the code
- **Linting**
  - [x] There should be no linting errors or warnings in the `file-name.lint.txt` file
  - [x] Spelling should be correct in all code, comments and docstrings
  - [x] The code was linted just before pushing
- **The solution file**
  - [x] Has a JSDoc comment
  - [x] Has an exported function
  - [x] Does no other code or comments (you can use the sandbox for saving extra ideas)
- **The function body _Should_ ...**
  - [x] Be well and consistently formatted
  - [x] Have clear and helpful variable names
  - [x] Declare variables in the smallest scope possible
  - [x] Be a simple and readable implementation
  - [x] Use short comments to describe the strategy
  - [x] Use short comments to describe important language feature choices
- **The function body should _Not_ ...**
  - [ ] Have extra comments that are not about strategy or important decisions
  - [x] Have any user interactions (`prompt`, `alert`, `confirm`)
  - [x] Have any `console` calls (`.log`, `.assert`, ...)
  - [x] Have any `debugger` statements

### PR DESCRIPTION
**Pull request by rago89**
---
name: solution code review
about: PR for code 
title: Intersection: Creates an array of values that are in two arrays
---

Before the reviewer merges this PR to main/master, all of these boxes must be checked. Only the reviewer can check boxes.

## Behavior

> What the function does - inputs and outputs

- **The Function**
  - [x] The function's name is clear and helpful
  - [x] The parameter names are clear and helpful
  - [x] Default parameters are used if possible
- **JSDoc Comment**
  - [x] The docstring is well-formatted
  - [x] There are no spelling mistakes
  - [x] The description of the function's behavior is clear
  - [x] All parameters are documented with types
  - [x] Documented parameter names match the function's parameter names
  - [x] Parameters have a good description
  - [x] If default parameters are used, they are the correct type
  - [x] The return value is documented with a type
  - [x] The return value has a clear description
- **README**
  - [x] The documentation is rendered into the README
  - [x] Docs were generated just before pushing
  - [x] Docs in the README matches the JSDoc comment
  - [x] The README contains at least one use case
- **Unit Tests**
  - [x] Unit tests are present
  - [x] There are at least 8 test cases
  - [x] All `describe`s and `it`s have clear and helpful names
  - [x] All test names actually describe what is being tested
  - [x] Tests are grouped logically using `describe`
  - [x] There is only one `expect` inside every `it`
  - [x] There is no logic (loops, conditionals, ...) in the test cases
  - [x] Side-effects are tested, if necessary
  - [x] Tests were run just before pushing
  - [x] All tests pass (`file-name.pass.txt`)
  - [x] Tricky edge cases are tested (extra credit)

## Strategy

> The logic behind the solution

- **The Function**
  - [x] Variables have names that help understand the strategy
  - [x] Spacing and formatting help to understand the strategy
  - [x] The code is written as simply and clearly as possible
  - [ ] Comments _are_ used to label important goals in your strategy
  - [x] Comments _are not_ used to describe how the code works
- **README**
  - [x] There is an explanation of the solution's strategy in the README
  - [x] The explanation is clear and understandable
  - [x] The explanation actually matches the function that was written

## Implementation

> The code used to write the solution

- **Formatting**
  - [x] The code was formatted just before pushing
  - [x] No errors occurred while formatting the code
- **Linting**
  - [x] There should be no linting errors or warnings in the `file-name.lint.txt` file
  - [x] Spelling should be correct in all code, comments and docstrings
  - [x] The code was linted just before pushing
- **The solution file**
  - [x] Has a JSDoc comment
  - [x] Has an exported function
  - [x] Does no other code or comments (you can use the sandbox for saving extra ideas)
- **The function body _Should_ ...**
  - [x] Be well and consistently formatted
  - [x] Have clear and helpful variable names
  - [x] Declare variables in the smallest scope possible
  - [x] Be a simple and readable implementation
  - [x] Use short comments to describe the strategy
  - [x] Use short comments to describe important language feature choices
- **The function body should _Not_ ...**
  - [x] Have extra comments that are not about strategy or important decisions
  - [x] Have any user interactions (`prompt`, `alert`, `confirm`)
  - [x] Have any `console` calls (`.log`, `.assert`, ...)
  - [x] Have any `debugger` statements

### CODE REVIEW

**Comment by erkamguresen** [comment]
Check list is unchecked.

**Comment by rago89** [comment]
> Check list is unchecked. 

I think that the checklist is for the reviewer

**Comment by erkamguresen** [comment]
> > Check list is unchecked.
> 
> I think that the checklist is for the reviewer

Yes I looked it again and it says 
'Before the reviewer merges this PR to main/master, all of these boxes must be checked. Only the reviewer can check boxes.'

The reason I write that was in class video 1.15.20 - 1.16.30 Thibault said a checklist before the pull request. 

and somethings on the checklist can not easily check by the reviewer  like 

The code was formatted just before pushing
The code was linted just before pushing

File: src/intersection/intersection.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+/**
+ * Creates an array of values that are in both the first and the second arrays.
+ *
+ * @param {Array} [array=[]] - The array to inspect.
+ * @param {Array} [values=[]] - The values to exclude.
+ * @returns {Array} Returns the new array of filtered values.
```
Comment (erkamguresen) [inline src/intersection/intersection.js]: 'filter' is not easy to understand and arises question what is filtered? And filter is not mentioned in any other explanation in JSdoc.  instead 'intersected'.

File: src/intersection/intersection.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+/**
+ * Creates an array of values that are in both the first and the second arrays.
+ *
+ * @param {Array} [array=[]] - The array to inspect.
+ * @param {Array} [values=[]] - The values to exclude.
```
Comment (erkamguresen) [inline src/intersection/intersection.js]: 'exclude' is causing misunderstandings.  If values of the second array is excluded, how the function is intersection? Can be simple as :
* @param {Array} [array=[]] - first array to inspect. 
*@param {Array} [values=[]] - second array to inspect..

File: src/intersection/intersection.spec.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
+import { arrayIntersection } from './intersection.js';
+
+describe('Intersect the elements that are equal and return them in a new array', () => {
```
Comment (erkamguresen) [inline src/intersection/intersection.spec.js]: describe part names actually describe what is being tested. Clear: OK.

File: src/intersection/intersection.spec.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  describe('testing default parameters', () => {
+    it('Passing an object literal', () => {
```
Comment (erkamguresen) [inline src/intersection/intersection.spec.js]: however 'it' part names have problems. 'It' part should describe what is the expected behavior of the code, not what is passed. For more clear examples : 

https://bignerdranch.com/blog/why-do-javascript-test-frameworks-use-describe-and-beforeeach/

and

https://mochajs.org/

An good example form last link:

You can read the code like poem

```Javascript
var assert = require('assert');
describe('Array', function() {
  describe('#indexOf()', function() {
    it('should return -1 when the value is not present', function() {
      assert.equal([1, 2, 3].indexOf(4), -1);
    });
  });
});

File: src/intersection/intersection.spec.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+      expect(arrayIntersection([undefined], [null])).toEqual([]);
+    });
+    it('strings', () => {
+      expect(arrayIntersection(['hello', ' bye'], ['hello'])).toEqual(['hello']);
+    });
+    it('all primitive', () => {
+      expect(arrayIntersection(['hello', 12, true, null, false, undefined], ['hello', true, 12, 25, 'bye'])).toEqual(['hello', 12, true]);
+    });
+  });
+});
```
Comment (erkamguresen) [inline src/intersection/intersection.spec.js]: Side effects are not tested. this can be like

```JS
    it('strings', () => {
      const firstArray =['hello', ' bye'];
      const secondArray =['hello'];
      arrayIntersection(firstArray, secondArray);
      expect(firstArray).toEqual(['hello', ' bye']);
    });
    it('strings', () => {
      const firstArray =['hello', ' bye'];
      const secondArray =['hello'];
      arrayIntersection(firstArray, secondArray);
      expect(secondArray).toEqual(['hello']);
    });

File: src/intersection/README.md [hunk 6]
```diff
@@ -1,34 +1,82 @@
...
+> Docs generated: Tue May 18 2021, 11:09:46 AM
+
+<!-- END DOCS -->
 
 ---
 
-## Docstring
+## Strategy
+
+I chose to used built-in array and object methods to solve this challenge, because once you know them they are easy to read and save a lot of lines and time.
```
Comment (erkamguresen) [inline src/intersection/README.md]: Language does not look professional, instead of using first or second person, using passive voice looks more professional.

File: src/intersection/README.md [hunk 7]
```diff
@@ -1,34 +1,82 @@
...
+
+---
+
+## Implementation
+
+I use an `if` and `else` statement:
```
Comment (erkamguresen) [inline src/intersection/README.md]: Language does not look professional, instead of using first or second person, using passive voice looks more professional.

**Review by erkamguresen — CHANGES_REQUESTED** [review]
- Evan notified me in the group chat as follows:
'I also just saw your difference.  Can you move this to a folder within the /difference folder like with the /sort-numbers example?  this way others can still submit PRs with different solutions to the difference challenge'
so can you move all your files into a folder and leave the original README file as it is in the main intersection folder?

and for the solution: 
The function body Should ...
 [ ] Use short comments to describe the strategy
 [ ] Use short comments to describe important language feature choices

**Comment by rago89** [comment]
Hello @Erkan thank you for checking the code, can you take a look again, I made some changes according your description

**Comment by rago89** [comment]
> * Evan notified me in the group chat as follows:
>   'I also just saw your difference.  Can you move this to a folder within the /difference folder like with the /sort-numbers example?  this way others can still submit PRs with different solutions to the difference challenge'
>   so can you move all your files into a folder and leave the original README file as it is in the main intersection folder?
> 
> and for the solution:
> The function body Should ...
> [ ] Use short comments to describe the strategy
> [ ] Use short comments to describe important language feature choices

**Comment by rago89** [comment]
> > * Evan notified me in the group chat as follows:
> >   'I also just saw your difference.  Can you move this to a folder within the /difference folder like with the /sort-numbers example?  this way others can still submit PRs with different solutions to the difference challenge'
> >   so can you move all your files into a folder and leave the original README file as it is in the main intersection folder?
> > 
> > and for the solution:
> > The function body Should ...
> > [ ] Use short comments to describe the strategy
> > [ ] Use short comments to describe important language feature choices

More comments added

**Comment by rago89** [comment]
Add one more comment to the solution

File: src/intersection/first-solution/intersection.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
+/**
+ * Creates an array of values that are in both the first and the second arrays.
+ *
+ * @param {Array} [firstArr=[]] - First array to be compared for repeated values.
+ * @param {Array} [secondArr=[]] - Second array to be compared for repeated values.
+ * @returns {Array} Returns the new array with only intersected values.
+ */
```
Comment (erkamguresen) [inline src/intersection/first-solution/intersection.js:7]: If you throw exception/error you must explain it in JS doc.  
https://jsdoc.app/tags-throws.html

Otherwise users will not know that the function throws stg.

File: src/intersection/first-solution/intersection.js [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+export const arrayIntersection = (firstArr = [], secondArr = []) => {
```
Comment (erkamguresen) [inline src/intersection/first-solution/intersection.js:9]: If you are going to throw error when no values are passed, what is the use of initializing them?

File: src/intersection/first-solution/intersection.js [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+  let newArrayIntersected = [];
+  // Test if the arguments are arrays and make sure that arguments are passed
+  if (!Array.isArray(firstArr) || !Array.isArray(secondArr)) {
+    throw new TypeError('The arguments must be arrays');
+  } else if (!firstArr.length || !secondArr.length) {
+    throw new Error('Must pas two arrays to intercept the values');
```
Comment (erkamguresen) [inline src/intersection/first-solution/intersection.js:14]: Mathematically empty set can intersect. 
https://proofwiki.org/wiki/Intersection_with_Empty_Set
https://en.wikipedia.org/wiki/Empty_set

think about a checkout in an online shop. image that I can have coupons which do discount on certain products. Suppose that I have no available coupon at the moment. And have 2 chairs in my basket. I press checkout to buy the items in my basket. If you throw error whole checkout will crash/fail. But if you return empty none of the items will be discounted and I would continue. May be the example is not %100 fit but I hope the reasoning is understandable.

File: src/intersection/first-solution/intersection.js [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+  } else {
+    // test if arrays have the same values, return only the values matched
+    newArrayIntersected = [...new Set(firstArr.filter((x) => secondArr.includes(x)))];
```
Comment (erkamguresen) [inline src/intersection/first-solution/intersection.js]: Why initialize the variable newArrayIntersected, since you never do an adding operation (push, shift etc..). Here you are creating a new Set and overriding the pointer. There is no way that this function will return the initialized array in the memory. This way it uses unnecessary heap space. 
https://medium.com/@ethannam/javascripts-memory-model-7c972cd2c239

Comment (erkamguresen) [reply src/intersection/first-solution/intersection.js]: With this approach heap is used 3 times, 
1 line 10
2 new set
3 [...
Solution can use heap once with simple methods

File: src/intersection/first-solution/intersection.js (hunk 10 above)
Comment (rago89) [reply src/intersection/first-solution/intersection.js:14]: the purpose of the function is to intersect values and in order to do that, I  have create the code to work only for intersection, if one argument is missing it will not work, of course I could make it more flexible but, that is according of the behavior that you want it to have

File: src/intersection/first-solution/intersection.js (hunk 11 above)
Comment (erkamguresen) [inline src/intersection/first-solution/intersection.js]: No need to use set since filter method returns a new array:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter

So no need to convert array to set then again to array. Unnecessary use of cpu.

You can visually see and play with it here:
http://www.pythontutor.com/live.html#code=const%20firstArr%20%3D%20%5B1,2%5D%3B%0Aconst%20secondArr%20%3D%20%5B2,3%5D%3B%0A%0Aconst%20theFilteredArray%20%3D%20firstArr.filter%28%28x%29%20%3D%3E%20secondArr.includes%28x%29%29%3B%0A%0AfirstArr%5B1%5D%20%3D%200%3B%0AsecondArr%5B0%5D%20%3D%200%3B%0A%0AtheFilteredArray.push%285%29%3B%0A%0Aconst%20theSet%20%3D%20new%20Set%28theFilteredArray%29%3B%0A%0A%0A%0Aconst%20newArrayIntersected%20%3D%20%5B...theSet%5D%3B&cumulative=false&curInstr=13&heapPrimitives=nevernest&mode=display&origin=opt-live.js&py=js&rawInputLstJSON=%5B%5D&textReferences=false

File: src/intersection/first-solution/intersection.js [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+ * @param {Array} [firstArr=[]] - First array to be compared for repeated values.
+ * @param {Array} [secondArr=[]] - Second array to be compared for repeated values.
+ * @returns {Array} Returns the new array with only intersected values.
+ */
+
+export const arrayIntersection = (firstArr = [], secondArr = []) => {
+  let newArrayIntersected = [];
+  // Test if the arguments are arrays and make sure that arguments are passed
+  if (!Array.isArray(firstArr) || !Array.isArray(secondArr)) {
+    throw new TypeError('The arguments must be arrays');
```
Comment (erkamguresen) [inline src/intersection/first-solution/intersection.js:12]: Just to point the importance of @throw in JSDoc  that all calls of this function will required to be in try-catch blocs otherwise when you throw error the program will stop:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/throw 

"The throw statement throws a user-defined exception. Execution of the current function will stop (the statements after throw won't be executed), and control will be passed to the first catch block in the call stack. If no catch block exists among caller functions, the program will terminate."

File: src/intersection/first-solution/intersection.js [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+  } else if (!firstArr.length || !secondArr.length) {
+    throw new Error('Must pas two arrays to intercept the values');
+  } else {
+    // test if arrays have the same values, return only the values matched
+    newArrayIntersected = [...new Set(firstArr.filter((x) => secondArr.includes(x)))];
+  }
+  return newArrayIntersected;
```
Comment (erkamguresen) [inline src/intersection/first-solution/intersection.js]: will this work? 
delete line 10;
delete line 16 else part:
delete line 18:
line 20: return firstArr.filter((x) => secondArr.includes(x));

File: src/intersection/intersection.spec.js (hunk 4 above)
Comment (colevandersWands) [reply src/intersection/intersection.spec.js]: @erkamguresen . oops, looks like I've a lot of example tests to rewrite in the module!

File: src/intersection/first-solution/intersection.js (hunk 11 above)
Comment (rago89) [reply src/intersection/first-solution/intersection.js]: True, thanks for note this, there is no need of new variable, filter does not modify the array instead it returns a new array, but I keep the `else` statement because maintains readability  to the control structure.  Changes pushed

Comment (rago89) [reply src/intersection/first-solution/intersection.js]: Is exactly the same, the difference is that you expand the code in several lines to understand what I did which is good, if you read my implementation, is well explained that line of code

File: src/intersection/first-solution/intersection.js (hunk 12 above)
Comment (rago89) [reply src/intersection/first-solution/intersection.js:12]: That is a documentation explicit to explain how the throw works, and in mine I'm explaining how intersection works

File: src/intersection/first-solution/intersection.js (hunk 13 above)
Comment (rago89) [reply src/intersection/first-solution/intersection.js]: There are many several ways to refactor a code, all are valid, even in yours I can make it in one line of code and may be more readable, but that doesn't mean that yours isn't valid or less readable, accept suggestions can be valid according to your criteria, you can see some of your suggestions adapted in my code, you don't have to expect that I write the code exactly as you want, your role is to check that everything is ok, and of course make suggestions to improve the code.
